pkg/lib/catalogsource/image_template.go

```diff
@@ -0,0 +1,356 @@
+// of a value in the slice means that the template was either not found in the cache or its value has not been
+// fetched yet). Providing an empty catalogImageTemplate results in empty processedCatalogImageTemplate and
+// zero length unresolvedTemplates
+func ReplaceTemplates(catalogImageTemplate string) (processedCatalogImageTemplate string, unresolvedTemplates []string) {
+
+	// init to empty slice
+	unresolvedTemplates = []string{}
+
+	// templateReplacer function determines the replacement value for the given template
+	var templateReplacer = func(template string) string {
+		replacement, ok := templateNameToReplacementValuesMap[template]
+		if ok {
+			// found a template, but check if the value is missing and keep record of
+			// what templates were unresolved
+			if replacement == templateMissing {
+				unresolvedTemplates = append(unresolvedTemplates, template)
+			}
+			return replacement
+		} else {
+			// probably should not get here, but in case there is no match,
+			// just return template unchanged, but keep record of
+			// what templates were unresolved
+			unresolvedTemplates = append(unresolvedTemplates, template)
+			return template
+		}
+	}
+
+	// if image is present, perform template substitution
+	if catalogImageTemplate != "" {
+		processedCatalogImageTemplate = regexImageTemplates.ReplaceAllStringFunc(catalogImageTemplate, templateReplacer)
+	}
+	return
+}
+
+// GetCatalogTemplateAnnotation is a helper function to extract the catalog image template annotation.
+// Returns empty string if value not set, otherwise returns annotation.
+func GetCatalogTemplateAnnotation(catalogSource *v1alpha1.CatalogSource) string {
+	if catalogSource == nil {
+		return ""
+	}
+	if catalogImageTemplate, ok := catalogSource.GetAnnotations()[CatalogImageTemplateAnnotation]; !ok {
+		return ""
+	} else {
+		return catalogImageTemplate
+	}
+}
+
+// GVK_Key uniquely represents a Group/Version/Kind (with optional name/namespace)
+// and can be used as a key for retrieval of data associated with this key
+type GVK_Key struct {
+	schema.GroupVersionKind
+	name      string
+	namespace string
+}
+
+func InitializeCatalogSourceTemplates(catalogSource *v1alpha1.CatalogSource) []GVK_Key {
+
+	// capture a list of keys that were found in this catalog source
+	foundGVKs := []GVK_Key{}
+
+	// findNamedMatches will return a map whose key is the named capture group, and value is the value of the capture group
+	findNamedMatches := func(str string) map[string]string {
+		// Note: matches and names indices are "in sync"
+		matches := regexGVKTemplate.FindStringSubmatch(str)
+		names := regexGVKTemplate.SubexpNames()
+
+		results := map[string]string{}
+		for i, match := range matches {
+			// only add named groups to the map
+			if names[i] != "" {
+				results[names[i]] = match
+			}
+		}
+		return results
+	}
+
+	catalogImageTemplate := GetCatalogTemplateAnnotation(catalogSource)
+	if catalogImageTemplate != "" {
```

Comment:
you could return early `if catalogImageTemplate == "  {...` instead of nesting this large block

Reply:
fixed